# Lab 02 - Hooke's Law

In [5]:
import numpy as np
import data_entry

Edit this cell and provide the information requested:

- Your name:
- Your student number:
- Partner's name:


## Part A - Introduction

***TA grading feedback here:***
- (Some notes about TA feedback)

Review of Hooke's Law:

- Approximation describing how springs/elastic objects deform in response to an applied force. 
- Accurate for small deviations
- Useful for modelling physical systems

The spring constant ($k$) is the proportionality constant that relates the change in force ($\Delta F$) needed to extend or compress a spring by some distance ($\Delta x$)

$$ k = \frac{\lvert \Delta F \rvert}{\lvert \Delta x \rvert} = g \frac{\lvert m_2 - m_1 \rvert}{\lvert x_2 - x_1 \rvert}$$

$x_1, m_1$: uncompressed position of the top of the spring and mass

$x_2, m_2$: compressed position of the top of the spring and mass

Experimental goal:
See if deviation from Hooke's Law can be detected by comparing large vs. small $\Delta x = x_2 - x_1$.

## Part B - Uncertainty Discussions

***Include notes on class discussion here:***
- Report uncertainty with 2 sig figs, and measurement has same decimal places as uncertainty.
- Rectangular uncertainty is half-width divided by $\sqrt{3}$
- u[m2] should be larger than u[m1] because of flucatuations, and should be Gaussian not rectangular

Relative uncertainty: Given $A \pm u[A]$, the relative uncertainty is given as:
    
$$ \frac{u[A]}{A} $$

More precise measurement = smaller relative uncertainty

Relative uncertainty in k is given by:

$$\frac{u[k]}{k}=\sqrt{\left(\frac{u[\Delta x]}{\Delta x}\right)^{2}+\left(\frac{u[\Delta m]}{\Delta m}\right)^{2}}$$

Can be reduced by:
- Decreasing $u[\Delta x]$, $u[\Delta m]$
- Increasing $\Delta x, \Delta m$

## Part C - Invention Activity

***Make sure to include a formula for the new statistical tool and a description of how to use and interpret it.***


Metric to evaluate whether two measurements with uncertainties are distinguishable.

$$t' = \frac{\lvert A - B\rvert}{\sqrt{(u[A])^2 + (u[B])^2}}$$

- If $t' \geq 3$, reasonably confident that the values are different. However, do a better measurement to increase your confidence further.
- If $1 < t' < 3$, values in tension; do another measurement to resolve.
- If $t' \leq 1$, Not at all confident they are different, but more precise measurement may uncover hidden difference.

## Part D - First round of measurements

***Include your initial measurement plan in the Markdown cell below, then follow the instructions on setting up your data spreadsheet.***

Goal: For this week’s experiment, we are investigating the limitations of Hooke’s Law by measuring the spring “constant” for different ranges of compression of the spring. 

Measurement plan:

1. Put the spring on the scale and TARE, measure the uncompressed length, and record the uncompressed mass reading (should be 0).
2. Push down spring around 3cm. There is some fluctuation, so take the average/center value observed for the length and scale value.
3. Repeat step 2, compressing spring down 7cm.

Uncertainties:

- Uncertainty in $x_1$ is Gaussian because human error in reading off length value... Determined by estimating 95% CI as boundaries of reasonable values, dividing by 4.
- Uncertainty in $m_1$ is Rectangular because digital measurement... Determined by dividing most precise decimal place in half and then by sqrt(3)
- Uncertainty in $x_2$ is Gaussian because human error + fluctuations in hand... Determined by estimating 95% CI as boundaries of reasonable length readings from unstable hand and dividing by 4
- Uncertainty in $m_2$ is Gaussian because fluctuations in the reading... Determined by estimating 95% CI as min/max values on scale seen and dividing by 4.

Some calculations for uncertainties:

In [6]:
ux1 = 0.004/4
print(ux1)

ux2small = 0.006/4
print(ux2small)

ux2large = 0.004/4
print(ux2large)

um1 = 0.001/2/np.sqrt(3)
print(um1)

um2large = 0.03/4
print(um2large)

um2small = 0.05/4
print(um2small)

0.001
0.0015
0.001
0.0002886751345948129
0.0075
0.0125


Uncomment the code below (i.e., remove the '#') to set up your initial spreadsheet 'lab02_data_round1.csv'. Any edits you make will automatically be saved. When you are finished inputting your data, press 'Generate Vectors' to transform each column into a Python vector. For instance, if you title a column 'm1', then a Python vector called 'm1Vec' will be generated. The idea is to use these generated vectors to perform the necessary calculations to obtain $k$ and $u[k]$ for small and large compressions.

In [8]:
de_round1 = data_entry.sheet("lab02_data_round1")

Sheet name: lab02_data_round1.csv


In [10]:
g = 9.801 # in Vancouver

Calculating k:

In [11]:
kVec = g*np.abs(m2Vec-m1Vec)/np.abs(x2Vec-x1Vec)
print("Spring constant k = {}".format(kVec))

Spring constant k = [141.7878     137.59096154]


Calculating uncertainty in k:

In [12]:
DxVec = x2Vec - x1Vec # delta x
DmVec = m2Vec - m1Vec # delta m

uDxVec = np.sqrt(ux1Vec**2 + ux2Vec**2) # uncertainty in delta x 
uDmVec = np.sqrt(um1Vec**2 + um2Vec**2) # uncertainty in delta m

ukVec = kVec*np.sqrt((uDxVec/DxVec)**2 + (uDmVec/DmVec)**2) # uncertainty in k
print("Spring constant uncertainty u[k] = {}".format(ukVec))

Spring constant uncertainty u[k] = [17.73241145  4.42223501]


Relative uncertainties:

In [13]:
print("Spring constant relative uncertainty u[k]/k = {}".format(ukVec/kVec))

Spring constant relative uncertainty u[k]/k = [0.12506303 0.03214045]


t' score:

In [14]:
ksmall = kVec[0]
klarge = kVec[1]
uksmall = ukVec[0]
uklarge = ukVec[1]

t = np.abs(ksmall - klarge)/np.sqrt(uksmall**2 + uklarge**2)
print("t-score = {}".format(t))

t-score = 0.22964264738391663


So for the small compression we have $k = 142 \pm 18 N/m$ and for the large compression we have $k = 137.6 \pm 4.4 N/m$. The relative uncertainties are 0.13 and 0.032 respectively. The t-score between these two values is 0.23, which is less than 1 so we are not at all confident that these two values are different.

## Part E - Group Discussion

***Comparisons can be measurement techniques and/or measured/calculated values. Don't forget to provide the names of the group members you compare to.***

a) (Some comparison with other groups of values...)

b) (Some comparison of relative uncertainties... Comparison of the numerical values by writing them down, why certain groups may have had lower/higher uncertainty values.) 

c) Looking at the t-score, we are not at all confident that the two values are different as t is much less than 1.

d) After discussing, we looked at the expression for the relative uncertainty in k and saw that it is best if we maximize $\Delta m$ and $\Delta x$, so we will take $\Delta x$ for the large compression to be 9cm (the largest) and $\Delta x$ for the small compression to be 5cm (the largest) so as to minimize the relative uncertainty. We saw that groups that did this had lower uncertainties so we conjecture that this will help to reduce ours. Also, instead of using our hands, we will also make towers of books and press down on the springs in a more stable way using a ruler so as to minimize the fluctuations (and hence $u[\Delta x], u[\Delta m]$) that arise from our unstable pressing down on the spring.

## Part F - Cycles of Improvement

Uncomment the code below to set up a new spreadsheet for your improvement round data. This code will copy your old spreadsheet so that all you have to do is overwrite the data from your initial round with your improvement round data. *IMPORTANT NOTE: do not run the code below until you are finished with your data entry in Part D.* 

In [16]:
de_round2 = data_entry.sheet_copy("lab02_data_round1", "lab02_data_round2")

Copied lab02_data_round1.csv to lab02_data_round2.csv
Creating undo file
Sheet name: lab02_data_round2.csv


Calculating k:

In [17]:
kVec = g*np.abs(m2Vec-m1Vec)/np.abs(x2Vec-x1Vec)
print("Spring constant k = {}".format(kVec))

Spring constant k = [139.1742 140.481 ]


Calculating uncertainty in k:

In [18]:
DxVec = x2Vec - x1Vec # delta x
DmVec = m2Vec - m1Vec # delta m

uDxVec = np.sqrt(ux1Vec**2 + ux2Vec**2) # uncertainty in delta x 
uDmVec = np.sqrt(um1Vec**2 + um2Vec**2) # uncertainty in delta m

ukVec = kVec*np.sqrt((uDxVec/DxVec)**2 + (uDmVec/DmVec)**2) # uncertainty in k
print("Spring constant uncertainty u[k] = {}".format(ukVec))

Spring constant uncertainty u[k] = [4.05701423 2.27382633]


Relative uncertainties:

In [19]:
print("Spring constant relative uncertainty u[k]/k = {}".format(ukVec/kVec))

Spring constant relative uncertainty u[k]/k = [0.02915062 0.01618601]


t' score:

In [20]:
ksmall = kVec[0]
klarge = kVec[1]
uksmall = ukVec[0]
uklarge = ukVec[1]

t = np.abs(ksmall - klarge)/np.sqrt(uksmall**2 + uklarge**2)
print("t-score = {}".format(t))

t-score = 0.2809858004403343


From our improvement round, we now have $k=139.2 \pm 4.1 N/m$ for the small compression and $k=140.5 \pm 2.3 N/m$ for the large compression.  With these improved measurements, our t-score is $0.28$ which is still less than one. Therefore, we are still not at all confident the large and small compressions have different $k$. In the future, we could do even better measurements to see if there is a very small hidden difference to uncover.

## Summary of Results

***Tips:***
- A brief summary of how your key results and the results of your comparisons evolved throughout the lab.
- 1-2 sentences to describe and justify your major improvements and how they proved to be successful or unsuccessful. It may be helpful to use relative uncertainty to quantify your argument. 
- 1-2 sentences addressing the experimental goal and describing the conclusions you can make from the results of your comparisons. 
- 1 sentence containing your thoughts for further investigation. If you were to spend more time with this experiment, what would you plan to do next?

From our best round of measurements, we found that $k=139.2 \pm 4.1 N/m$ for the small compression and $k=140.5 \pm 2.3 N/m$ for the large compression. These values are similar to our initial round of measurements, except our relative uncertainty has decreased from $0.125$ to $0.029$ (small compression) and $0.032$ to $0.016$ (large compression). This tells us that our improvements of using maximum compressions within the allowed ranges (to directly lower uncertainty based on the $u[k]/k$ formula), as well as stabilizing the compression by using an external "tower" of books (to reduce $u[\Delta x]$ and especially $u[\Delta m]$) were successful. Taking a t-score between the two results, we find $t' = 0.28$ and so we are not at all confident the two $k$ are different because $t'<1$. From the results of our experiment, we would conclude that Hooke's law holds across the two ranges.